### PR TITLE
Use case-insensitive header keys for http probes

### LINKED ISSUE #44291: localhost HOST header gives 400 bad Request for the health check
### Bug Description

Our system use Django framework, so for the allowed_hosts we need localhost host header for the health check from the kubelet. To achieve it we use httpHeaders for the livenessProbe and readinessProbe.
```
        readinessProbe:
          failureThreshold: 10
          httpGet:
            httpHeaders:
            - name: HOST
              value: localhost
            path: /status/
            port: 8000
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 30
```

As I checked, if I use `httpHeaders` to push localhost host header, I found out that the istio-proxy gives 400 bad request like below.
```
00:08:14.181896 IP 10.119.26.76.37666 > 10.119.28.178.15020: Flags [P.], seq 1:155, ack 1, win 491, options [nop,nop,TS val 3656328203 ecr 791935060], length 154
E...JN@.....
w.L
w...":.'..+..;.....L......
..$./3.TGET /app-health/service-name/readyz HTTP/1.1
Host: 10.119.28.178:15020
User-Agent: kube-probe/1.22+
Accept: */*
HOST: localhost
Connection: close


00:08:14.181909 IP 10.119.28.178.15020 > 10.119.26.76.37666: Flags [.], ack 155, win 489, options [nop,nop,TS val 791935060 ecr 3656328203], length 0
E..4.m@....j
w..
w.L:.."..;.'.......L......
/3.T..$.
00:08:14.182009 IP 10.119.28.178.15020 > 10.119.26.76.37666: Flags [P.], seq 1:104, ack 155, win 489, options [nop,nop,TS val 791935060 ecr 3656328203], length 103
E....n@.....
w..
w.L:.."..;.'.......Ly.....
/3.T..$.HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Connection: close
```

If I remove httpHeaders, it works fine. (also I put allow all from the allowed_hosts in Django)
```
00:10:20.636776 IP 10.119.26.76.35548 > 10.119.24.149.15020: Flags [P.], seq 1:138, ack 1, win 491, options [nop,nop,TS val 2930496383 ecr 2308459026], length 137
E.....@...1.
w.L
w....:.ag0dn.......H~.....
......J.GET /app-health/service-name/readyz HTTP/1.1
Host: 10.119.24.149:15020
User-Agent: kube-probe/1.22+
Accept: */*
Connection: close


00:10:20.636796 IP 10.119.24.149.15020 > 10.119.26.76.35548: Flags [.], ack 138, win 489, options [nop,nop,TS val 2308459026 ecr 2930496383], length 0
E..4t.@....P
w..
w.L:...n...ag0.....G......
..J.....
00:10:20.641938 IP 10.119.24.149.15020 > 10.119.26.76.35548: Flags [P.], seq 1:95, ack 138, win 489, options [nop,nop,TS val 2308459031 ecr 2930496383], length 94
E...t.@.....
w..
w.L:...n...ag0.....HS.....
..J.....HTTP/1.1 200 OK
Date: Fri, 07 Apr 2023 00:10:20 GMT
Content-Length: 0
Connection: close
```

Can I know the principle of the health check of the istio-proxy through `ISTIO_KUBE_APP_PROBERS` to troubleshoot together and why localhost host header influence to the health check?

Many thanks for your help in advance !!!

### Version

```prose
❯ istioctl version
client version: 1.16.3
control plane version: 1.16.3
data plane version: 1.16.3 (183 proxies)
```

### Additional Information

_No response_

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### PR DESCRIPTION
**Pull request by dddddai**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
We should be using canonical format for http header keys, which is required by [golang http spec](https://github.com/golang/go/blob/go1.20.3/src/net/http/header.go#L22-L24). Otherwise it could generate invalid headers

#### Which issue(s) this PR fixes:

Fixes https://github.com/istio/istio/issues/44291

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by pacoxu — APPROVED** [review]
/lgtm

File: pkg/probe/http/request.go [hunk 1]
```diff
@@ -113,7 +113,8 @@ func formatURL(scheme string, host string, port int, path string) *url.URL {
 func v1HeaderToHTTPHeader(headerList []v1.HTTPHeader) http.Header {
 	headers := make(http.Header)
 	for _, header := range headerList {
-		headers[header.Name] = append(headers[header.Name], header.Value)
+		key := http.CanonicalHeaderKey(header.Name)
+		headers[key] = append(headers[key], header.Value)
```
Comment (howardjohn) [inline pkg/probe/http/request.go:117]: I think you can do `headers.Add` to handle this as well

File: pkg/probe/http/request_test.go [hunk 2]
```diff
@@ -64,6 +64,17 @@ func Test_v1HeaderToHTTPHeader(t *testing.T) {
...
 			},
 		},
+		{
+			name: "case insensitive",
+			headerList: []v1.HTTPHeader{
+				{Name: "HOST", Value: "example.com"},
+				{Name: "FOO-bAR", Value: "value"},
+			},
+			want: http.Header{
+				"Host":    {"example.com"},
```
[the remark this answers is not in the document]
Comment (dddddai) [reply pkg/probe/http/request_test.go:74]: > Or do you still get 2 Host values if they also set the explicit `host` field separate from headers

No, the `Host` field is used to specify the url
https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/pkg/probe/http/request.go#L46-L57

File: pkg/probe/http/request.go (hunk 1 above)
Comment (dddddai) [reply pkg/probe/http/request.go]: Updated

File: pkg/probe/http/request.go [hunk 3]
```diff
@@ -113,7 +113,7 @@ func formatURL(scheme string, host string, port int, path string) *url.URL {
 func v1HeaderToHTTPHeader(headerList []v1.HTTPHeader) http.Header {
 	headers := make(http.Header)
 	for _, header := range headerList {
-		headers[header.Name] = append(headers[header.Name], header.Value)
+		headers.Add(header.Name, header.Value)
```
Comment (pacoxu) [inline pkg/probe/http/request.go:116]:
```suggestion
		headers.Add(http.CanonicalHeaderKey(header.Name), header.Value)
```
I think this should be keeping using CanonicalHeaderKey here.

Comment (dddddai) [reply pkg/probe/http/request.go:116]: It's handled inside the `Add` method. The test is passed

Comment (pacoxu) [reply pkg/probe/http/request.go:116]:
```
func (h MIMEHeader) Add(key, value string) {
	key = CanonicalMIMEHeaderKey(key)
	h[key] = append(h[key], value)
}
```
Got it. Sorry for the noise.
/lgtm

File: pkg/probe/http/request_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aojea) [reply pkg/probe/http/request_test.go:74]: in golang the host field is generated from the request AFAIK https://github.com/golang/go/issues/29865#issuecomment-456305578

**Comment by dddddai** [comment]
> the golang comment seems an implementation detail of golang, or at least I couldn't find a place that says headers should be canonicalized (but I can be completely wrong here, feel free to correct me)

I don't think so, please see https://developer.mozilla.org/en-US/docs/web/http/headers
> An HTTP header consists of its **case-insensitive** name followed by a colon (:), then by its value.

**Comment by aojea** [comment]
so, yeah, header are case-insensitive https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive no point in mutating then

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by thockin** [comment]
It does seem like the HTTP spec says "case insensitive", which says to me that istio and Go are both wrong to demand canonical form.  That said, Postel's law makes sense here, and I don't see much harm in canonicalizing.  If someone was depending on case-sensitivity to have (e.g.) "Host" and "host" and "HOST" be differerent, they were already on the wrong side of the spec.

Perhaps we should also update the API comments to explain that `HTTPHeader.Name` will be canonicalized upon output, so case-variant names will be understood as the same header.

**Comment by aojea** [comment]
> which says to me that istio and Go are both wrong to demand canonical form.

that is not what happen, the istio problem is a mix of problems, the only thing is that by canonicalizing it gets solved , but is by chance

I don't see that Go demands canonical form, that sounds more like an internal thing, actually it respects the headers if you set them

**Comment by thockin** [comment]
Oh, I see the istio issue - are we setting the "Host" header somewhere, and not recognizing that "HOST" == "Host" ?

And, OK - Go doesn't *demand* canonical form, but it strongly suggests that they should be canonical, with an implied "or else..." 

So should we approve this PR, then, assuming the docs are updated?

**Comment by howardjohn** [comment]
> Oh, I see the istio issue - are we setting the "Host" header somewhere, and not recognizing that "HOST" == "Host" ?
> 
> And, OK - Go doesn't _demand_ canonical form, but it strongly suggests that they should be canonical, with an implied "or else..."
> 
> So should we approve this PR, then, assuming the docs are updated?

The Istio case isn't _really_ Istio specific. We rewrite the applications readiness probe to a port handled by the sidecar container, which is written in Go (this is terrible, I know). Go's HTTP stack rejects the request before it hits an istio code. The same thing would happen if you had any Go based readiness probe. It was reported as an Istio issue because the user's application _was_ some non-Go container that didn't mind the invalid request, so turning Istio on moved to Go-based probe which broke.

(This also uncovered another Istio issue where we duplicated headers on probes for ~similar reasons but that is mostly unrelated to this, if that was causing confusion).

You can reproduce this with the below deployment (no istio sidecar needed here, the app is Go-based):

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo
spec:
  selector:
    matchLabels:
      app: echo
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
      - name: echo
        image: gcr.io/istio-testing/app:latest
        readinessProbe:
          httpGet:
            httpHeaders:
            - name: HOST
              value: localhost
            path: /
            port: 8080
            scheme: HTTP
```

**Comment by thockin** [comment]
I'm trying to understand why the request is invalid - where is the "Host" (not "HOST") coming from?

**Comment by howardjohn** [comment]
> I'm trying to understand why the request is invalid - where is the "Host" (not "HOST") coming from?

I think Go must have some knowledge of `Host` but not other casings: https://go.dev/play/p/gjbESYvpbMl. Since the `Host` is a part of http.Request type

**Comment by thockin** [comment]
Ahh, of course.  That makes sense.

**Comment by howardjohn** [comment]
Looks like here: https://github.com/golang/go/blob/7c47a6b15782b13ecb76fd3c6c18e5f1edc34733/src/net/http/request.go#L1030

TBH might qualify as a Go bug as well...

**Comment by thockin** [comment]
@dddddai if you update the API docs (see https://github.com/kubernetes/kubernetes/pull/117182#issuecomment-1503659972) in staging/src/k8s.io/api/core/v1/types.go and pkg/apis/core/types.go then I will approve.

**Comment by dddddai** [comment]
Thanks for your review!
> I'm trying to understand why the request is invalid - where is the "Host" (not "HOST") coming from?

When sending requests, the Go http client [automatically adds a `Host` header](https://github.com/golang/go/blob/7c47a6b15782b13ecb76fd3c6c18e5f1edc34733/src/net/http/request.go#L629-630), which is derived from:

- [the `Host` field in Request](https://github.com/golang/go/blob/7c47a6b15782b13ecb76fd3c6c18e5f1edc34733/src/net/http/request.go#L238-L242)
- **OR** [`Url.Host`](https://github.com/golang/go/blob/7c47a6b15782b13ecb76fd3c6c18e5f1edc34733/src/net/http/request.go#L579-L584) , if the `Host` field is unspecified

For example:
```
      httpGet:
        httpHeaders:
        - name: HOST
          value: foo
        path: /status/
        port: 8000
        scheme: HTTP
```
kubelet _should_ have [set the `Host` field to “foo"](https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/pkg/probe/http/request.go#L84), however it will fail because we didn't canonicalize the key. As a result, the request will have 2 host headers:

- Host: <pod_ip> ([`Url.Host`](https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/pkg/probe/http/request.go#L57))
- HOST: foo (specified by the user)

**Comment by dddddai** [comment]
> this solves the problem on https://github.com/istio/istio/issues/44291 but , as a "fortunate" consequence

Why is it "fortunate"?

> You should get the same result if the configuration of the header in the probe is `Host` instead of `HOST`

Sorry I don't get it, could you explain why?

> I don't see that Go demands canonical form, that sounds more like an internal thing

Yes it is an internal thing. Canonical form is the way Go uses to meet the case-insensitive requirement. That's why we should be [using `headers.Add`](https://github.com/kubernetes/kubernetes/pull/117182/files#diff-57ea5c25f416bb939738fd23f2641f6911b077aaba01d160f1c1a7f536ef03faR116) to build headers. This makes sure everything is in canonical form, so that Go http client can correctly read/use them.
As you can see Go always [assumes headers are canonicalized](https://github.com/golang/go/blob/7c47a6b15782b13ecb76fd3c6c18e5f1edc34733/src/net/http/request.go#L635)

So I don't think this is a Go bug, it's just we are not using Go http client in a correct way?

**Comment by dddddai** [comment]
For another instance

      httpGet:
        httpHeaders:
        - name: USER-AGENT
          value: bar
        path: /status/
        port: 8000
        scheme: HTTP

Headers will also be messed if we don't canonicalize the key:
https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/pkg/probe/http/request.go#L72-L75

It will end up with
```
User-agent: kube-xxx
USER-AGENT: bar
```

**Comment by jinn-k** [comment]
> > Oh, I see the istio issue - are we setting the "Host" header somewhere, and not recognizing that "HOST" == "Host" ?
> > And, OK - Go doesn't _demand_ canonical form, but it strongly suggests that they should be canonical, with an implied "or else..."
> > So should we approve this PR, then, assuming the docs are updated?
> 
> The Istio case isn't _really_ Istio specific. We rewrite the applications readiness probe to a port handled by the sidecar container, which is written in Go (this is terrible, I know). Go's HTTP stack rejects the request before it hits an istio code. The same thing would happen if you had any Go based readiness probe. It was reported as an Istio issue because the user's application _was_ some non-Go container that didn't mind the invalid request, so turning Istio on moved to Go-based probe which broke.
> 
> (This also uncovered another Istio issue where we duplicated headers on probes for ~similar reasons but that is mostly unrelated to this, if that was causing confusion).
> 
> You can reproduce this with the below deployment (no istio sidecar needed here, the app is Go-based):
> 
> ```yaml
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: echo
> spec:
>   selector:
>     matchLabels:
>       app: echo
>   template:
>     metadata:
>       labels:
>         app: echo
>     spec:
>       containers:
>       - name: echo
>         image: gcr.io/istio-testing/app:latest
>         readinessProbe:
>           httpGet:
>             httpHeaders:
>             - name: HOST
>               value: localhost
>             path: /
>             port: 8080
>             scheme: HTTP
> ```
> 
> > If someone was depending on case-sensitivity to have (e.g.) "Host" and "host" and "HOST" be differerent, they were already on the wrong side of the spec.
> 
> I think this is a reasonable stance. I would note that Istio/Envoy does the same, which has caused _some_ complaints. But that is a very different scope, where we are proxying all their traffic vs just the probe, so its less likely to be an issue.

Hello @howardjohn 

I tried to test with the manifest, and when I queried the pod using the following command, it returned a 200 response code:
```
$ curl http://<pod ip>:8080/ -s -o /dev/null -D /dev/stdout -H 'HOST: localhost'
HTTP/1.1 200 OK
Date: Wed, 12 Apr 2023 02:55:13 GMT
Content-Length: 219
Content-Type: text/plain; charset=utf-8
```
I have a question that might be a bit of a newbie one, but I'm wondering if you could help me out. I'm trying to understand the process by which the kubelet receives a 400 response code.

I'm thinking that maybe I've misunderstood something, so any insight you can provide would be much appreciated.

Thanks!
```
# Pod describe log
Readiness probe failed: Get "http://10.119.27.10:8080/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Readiness probe failed: HTTP probe failed with statuscode: 400
```
Thank you.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Comment by dddddai** [comment]
Sorry I didn't update swagger docs

**Comment by aojea** [comment]
/lgtm

Thanks

**Comment by aojea** [comment]
> /hold
> 
> Sorry, I didn't see @aojea's comment.

my comment was addressed, thanks

/hold cancel

**Comment by thockin** [comment]
Should we backport this to 1.27 and 1.26 ?

**Comment by pacoxu** [comment]
> Should we backport this to 1.27 and 1.26 ?

@dddddai do you have time to do the cherry-pick? Refer to https://github.com/kubernetes/community/blob/f03b6d5692bd979f07dd472e7b6836b2dad0fd9b/contributors/devel/sig-node/cherry-picks.md?plain=1#L40-L42.

**Comment by dddddai** [comment]
OK, will submit tomorrow :)

**Comment by dddddai** [comment]
Yeah I think so

**Pull request merged**